File: NeurIPSScraper.java
package com.sarim.scraper;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.*;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Random;
import java.util.concurrent.*;

public class NeurIPSScraper {
    private static final String BASE_URL = "https://papers.nips.cc";
    private static final String STORAGE_DIRECTORY = "/Users/muhammadsarim/Desktop/NeurIPS_Papers-Java/";

    public static void main(String[] args) {
        try {
            prepareStorageDirectory();
            ExecutorService executorService = initializeExecutorService();
            CompletionService<Void> taskCompletionService = new ExecutorCompletionService<>(executorService);

            try (BufferedWriter csvWriter = initializeCSVWriter()) {
                for (int year = 2023; year >= 2019; year--) {
                    fetchAndProcessPapersForYear(year, csvWriter, taskCompletionService);
                }
            } catch (IOException e) {
                System.err.println("Error initializing the CSV writer.");
                e.printStackTrace();
            }

            shutDownExecutorService(executorService);
        } catch (IOException e) {
            System.err.println("Failed to create the storage directory.");
            e.printStackTrace();
        }
    }

    private static void prepareStorageDirectory() throws IOException {
        Files.createDirectories(Paths.get(STORAGE_DIRECTORY));
    }

    private static ExecutorService initializeExecutorService() {
        return Executors.newFixedThreadPool(7);
    }

    private static BufferedWriter initializeCSVWriter() throws IOException {
        BufferedWriter writer = new BufferedWriter(new FileWriter(STORAGE_DIRECTORY + "papers_output.csv"));
        writer.write("Year,Title,Authors,Paper URL,PDF URL\n");
        return writer;
    }

    private static void fetchAndProcessPapersForYear(int year, BufferedWriter writer, CompletionService<Void> completionService) {
        String yearPageUrl = BASE_URL + "/paper_files/paper/" + year;
        String yearFolder = STORAGE_DIRECTORY + year + "/";

        try {
            Files.createDirectories(Paths.get(yearFolder));
            Document yearPage = fetchDocumentWithDelay(yearPageUrl);

            Elements paperLinks = yearPage.select("ul.paper-list li.conference a");
            if (paperLinks.isEmpty()) {
                System.err.println("No papers found for year: " + year);
                return;
            }

            for (Element paperLink : paperLinks) {
                String paperTitle = paperLink.text().trim();
                String paperPageUrl = BASE_URL + paperLink.attr("href");
                completionService.submit(new PaperProcessingTask(writer, paperTitle, paperPageUrl, yearFolder, year));
            }
        } catch (IOException e) {
            System.err.println("Error fetching papers for year: " + year);
            e.printStackTrace();
        }
    }

    private static Document fetchDocumentWithDelay(String url) {
        try {
            Thread.sleep(1000 + new Random().nextInt(2000));
            return Jsoup.connect(url).get();
        } catch (IOException | InterruptedException e) {
            System.err.println("Failed to fetch document: " + url);
            e.printStackTrace();
            return null;
        }
    }

    private static void shutDownExecutorService(ExecutorService executorService) {
        try {
            executorService.shutdown();
            executorService.awaitTermination(1, TimeUnit.HOURS);
        } catch (InterruptedException e) {
            System.err.println("Executor service interrupted.");
            e.printStackTrace();
        }
    }

    static class PaperProcessingTask implements Callable<Void> {
        private final BufferedWriter csvWriter;
        private final String paperTitle;
        private final String paperPageUrl;
        private final String yearFolder;
        private final int year;

        public PaperProcessingTask(BufferedWriter writer, String paperTitle, String paperPageUrl, String yearFolder, int year) {
            this.csvWriter = writer;
            this.paperTitle = paperTitle;
            this.paperPageUrl = paperPageUrl;
            this.yearFolder = yearFolder;
            this.year = year;
        }

        @Override
        public Void call() throws InterruptedException {
            try {
                System.out.println("Processing paper: " + paperTitle);
                Document paperDocument = Jsoup.connect(paperPageUrl).get();

                String authors = extractAuthors(paperDocument);
                String pdfUrl = extractPdfUrl(paperDocument);

                if (!"N/A".equals(pdfUrl)) {
                    downloadPdf(pdfUrl, yearFolder, paperTitle);
                }

                savePaperDetailsToCSV(authors, pdfUrl);
            } catch (IOException e) {
                System.err.println("Error processing paper: " + paperTitle);
                e.printStackTrace();
            }

            return null;
        }

        private String extractAuthors(Document document) {
            Elements authorElements = document.select("i");
            StringBuilder authors = new StringBuilder();
            for (Element author : authorElements) {
                authors.append(author.text()).append("; ");
            }
            return authors.toString().trim();
        }

        private String extractPdfUrl(Document document) {
            Element pdfElement = document.selectFirst("a[href$=.pdf]");
            return (pdfElement != null) ? BASE_URL + pdfElement.attr("href") : "N/A";
        }

        private void downloadPdf(String pdfUrl, String savePath, String title) throws InterruptedException {
            String sanitizedTitle = title.replaceAll("[^a-zA-Z0-9]", "_") + ".pdf";
            String filePath = savePath + sanitizedTitle;
            System.out.println("Downloading PDF: " + pdfUrl);
            try {
                downloadFile(pdfUrl, filePath);
            } catch (IOException e) {
                System.err.println("Failed to download PDF: " + pdfUrl);
                e.printStackTrace();
            }
        }

        private void downloadFile(String url, String savePath) throws IOException {
            try (InputStream in = new URL(url).openStream(); FileOutputStream out = new FileOutputStream(savePath)) {
                byte[] buffer = new byte[1024];
                int bytesRead;
                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
                System.out.println("Downloaded: " + savePath);
            }
        }

        private void savePaperDetailsToCSV(String authors, String pdfUrl) throws IOException {
            synchronized (csvWriter) {
                csvWriter.write(String.format("%d,\"%s\",\"%s\",\"%s\",\"%s\"\n", year, paperTitle, authors, paperPageUrl, pdfUrl));
                csvWriter.flush();
            }
            System.out.println("Processed paper: " + paperTitle);
        }
    }
}

File: NeurIPSScraper.py
import os
import time
import random
import requests
from bs4 import BeautifulSoup
from concurrent.futures import ThreadPoolExecutor
import csv


BASE_URL = "https://papers.nips.cc"
STORAGE_DIRECTORY = "/Users/muhammadsarim/Desktop/NeurIPS_Papers-Python/"


def prepare_storage_directory():
    os.makedirs(STORAGE_DIRECTORY, exist_ok=True)


def fetch_and_process_papers_for_year(year, csv_writer):
    year_page_url = f"{BASE_URL}/paper_files/paper/{year}"
    year_folder = os.path.join(STORAGE_DIRECTORY, str(year))

    try:
        os.makedirs(year_folder, exist_ok=True)
        year_page = fetch_document_with_delay(year_page_url)

        paper_links = year_page.select("ul.paper-list li.conference a")
        if not paper_links:
            print(f"No papers found for year: {year}")
            return

        for paper_link in paper_links:
            paper_title = paper_link.text.strip()
            paper_page_url = BASE_URL + paper_link['href']
            process_paper(paper_title, paper_page_url, year_folder, year, csv_writer)

    except Exception as e:
        print(f"Error fetching papers for year: {year}")
        print(e)


def fetch_document_with_delay(url):
    try:
        time.sleep(1 + random.randint(0, 2))  # Simulating delay
        response = requests.get(url)
        response.raise_for_status()
        return BeautifulSoup(response.text, 'html.parser')
    except requests.exceptions.RequestException as e:
        print(f"Failed to fetch document: {url}")
        print(e)
        return None


def process_paper(paper_title, paper_page_url, year_folder, year, csv_writer):
    try:
        print(f"Processing paper: {paper_title}")
        paper_document = fetch_document_with_delay(paper_page_url)

        authors = extract_authors(paper_document)
        pdf_url = extract_pdf_url(paper_document)

        if pdf_url != "N/A":
            download_pdf(pdf_url, year_folder, paper_title)

        save_paper_details_to_csv(authors, pdf_url, paper_title, paper_page_url, year, csv_writer)

    except Exception as e:
        print(f"Error processing paper: {paper_title}")
        print(e)


def extract_authors(document):
    author_elements = document.select("i")
    authors = [author.text for author in author_elements]
    return "; ".join(authors)


def extract_pdf_url(document):
    pdf_element = document.select_one("a[href$=.pdf]")
    return BASE_URL + pdf_element['href'] if pdf_element else "N/A"


def download_pdf(pdf_url, save_path, title):
    sanitized_title = "".join(c if c.isalnum() else "_" for c in title) + ".pdf"
    file_path = os.path.join(save_path, sanitized_title)
    print(f"Downloading PDF: {pdf_url}")

    try:
        download_file(pdf_url, file_path)
    except Exception as e:
        print(f"Failed to download PDF: {pdf_url}")
        print(e)


def download_file(url, save_path):
    response = requests.get(url, stream=True)
    response.raise_for_status()
    
    with open(save_path, 'wb') as out_file:
        for chunk in response.iter_content(chunk_size=1024):
            if chunk:
                out_file.write(chunk)
    print(f"Downloaded: {save_path}")


def save_paper_details_to_csv(authors, pdf_url, paper_title, paper_page_url, year, csv_writer):
    try:
        csv_writer.writerow([year, paper_title, authors, paper_page_url, pdf_url])
        print(f"Processed paper: {paper_title}")
    except Exception as e:
        print(f"Error saving paper details for {paper_title}")
        print(e)


def main():
    try:
        prepare_storage_directory()
        with open(os.path.join(STORAGE_DIRECTORY, "papers_output.csv"), 'w', newline='', encoding='utf-8') as file:
            csv_writer = csv.writer(file)
            csv_writer.writerow(["Year", "Title", "Authors", "Paper URL", "PDF URL"])

            with ThreadPoolExecutor(max_workers=7) as executor:
                for year in range(2023, 2018, -1):
                    fetch_and_process_papers_for_year(year, csv_writer)

    except Exception as e:
        print("Error in main execution.")
        print(e)


if __name__ == "__main__":
    main()